citar-file.el

```diff
@@ -206,16 +206,21 @@ need to scan the contents of DIRS in this case."
                  (puthash key (nreverse filelist) files))
                files))))
 
-(defun citar-file--has-file-p (dirs extensions &optional additional-sep entry-field)
+(defun citar-file--make-file-predicate (dirs extensions &optional entry-field)
```

Comment:
Why this change in name?

The former seems more consistent to emacs predicate naming conventions?

Reply:
I wanted to make it clear that this function isn't a predicate, but that it *returns* a predicate. Usually the `foo-p` convention is used for the predicate functions themselves.

Reply:
Been a long day grading, but ...

Why the need to generate the predicates?

Wouldn't the resulting code be more clear to simply write them?

Reply:
This function does the job of scanning the library directories for all possible files and caching them in a hash table. The predicate it returns is very efficient, since it just checks that hash table instead of actually scanning the filesystem every time it's called.

Reply:
Ok,. but I'm talking about the example you gave; this:

``` elisp
(citar-select-ref :filter (citar-file--make-file-predicate citar-library-paths
                                                           citar-file-extensions
                                                           citar-file-variable))
```

... vs something like:

``` elisp
(citar-select-refs :filter #'citar-has-notes-p)
```

So can we somehownhabe both: the efficiency benefit + ease-of-use and clarity?

Reply:
Sure, I could add some small functions like:
```emacs-lisp
(defun citar--make-file-predicate ()
  (citar-file--make-file-predicate citar-library-paths
                                   citar-file-extensions
                                   citar-file-variable))

(defun citar--make-note-predicate ()
  (citar-file--make-file-predicate citar-notes-paths
                                   citar-file-note-extensions))
```
These functions would have to go in `citar.el`, rather than `citar-file.el`, because they use the `citar-library-paths` and `citar-notes-paths` variables. Then the example code would become
```emacs-lisp
(citar-select-refs :filter (citar-make-note-predicate))
```
In any case, the :filter function will have to be the result of a function call, rather than just the name of a function.

Reply:
I've pushed a new commit that adds these utility functions, and updated the PR opening comment to use them.

Reply:
I'm still not thrilled with the `make-*-predicste` names.

Maybe I'm misunderstanding some subtlety, but WDYT about simply `citar-note-filter`?

It seems more explanatory, and is consistent, for example, with the parser functions.

Reply:
How about `citar-has-file` and `citar-has-note`? I prefer these to `citar-note-filter` because that sounds like a function that performs filtering, like `seq-filter`.

I'm pushing a new commit with these names; it also updates the docstrings to give usage guidance.

And yes, this also works with `citar-select-refs`, since that's now just a wrapper calling `(citar-select-ref :multiple t ...)`.

Reply:
That works; merged!